provider/core.rb

```diff
@@ -184,11 +184,15 @@ def generate_objects(output_folder, types, version_name)
       version = @api.version_obj_or_default(version_name)
       @api.set_properties_based_on_version(version)
       (@api.objects || []).each do |object|
+        # exclude_if_not_in_version is destructive for 'object' and will remove
+        # properties that are not the required version.
+        excluded = object.exclude_if_not_in_version!(version)
+
         if !types.empty? && !types.include?(object.name)
           Google::LOGGER.info "Excluding #{object.name} per user request"
         elsif types.empty? && object.exclude
           Google::LOGGER.info "Excluding #{object.name} per API catalog"
-        elsif types.empty? && object.exclude_if_not_in_version(version)
+        elsif types.empty? && excluded
```

Comment:
```suggestion
        elsif types.empty? && object.excluded_at_version? version
```

It seems really silly to overload `exclude_if_not_in_version`/`exclude_if_not_in_version!` to both tell us whether a resource has been excluded and to remove the fields that are too high of a version.

What about splitting that method into an `excluded_at_version?` method as well as `exclude_fields_not_in_version!` (or similar naming) that we can call in the `else` block here instead of having to hoist it up?

Reply:
I definitely like that idea.

Reply:
K, I'll admit part of the reason I just did the rename was because I wasn't totally sure _why_ it was destructive. So after a little digging it looks like [this line](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/api/type.rb#L160) is the culprit that "removes" a deeply nested property. The subtlety is that line will change the property's `exclude` instance_variable from false to true, which combined with the fact that the .properties method rejects 'excluded' properties at read time means the property is no longer available.

However, the [top if check](https://github.com/GoogleCloudPlatform/magic-modules/pull/1260/files#diff-b6182235fa5ea9b0be8cadc36cd8b323R195) really only cares about the top level resource's exclude status, and the fact that the method recurses doesn't affect what it returns. Let me know if this makes sense, but I'll push up a change that should be a bit cleaner.